### PR TITLE
[18NY] More operation round functionality

### PR DESCRIPTION
**Pull request by crericha**
- Implement connection bonus tokens
- Place coal tokens with assignment
- Implement hex trains
- Implement tokens ($20 * shortest legal path from another station)

### CODE REVIEW

File: lib/engine/game/g_18_ny/game.rb [hunk 1]
```diff
@@ -383,6 +403,72 @@ def upgrade_cost(tile, _hex, entity, spender)
...
+          if (num_bonuses = route.stops.count { |stop| stop.hex.assigned?('connection_bonus') }).positive?
+            str += " + #{num_bonuses} Connection Bonus#{num_bonuses == 1 ? '' : 'es'}"
+          end
+
+          str
+        end
+
+        def routes_revenue(routes)
+          revenue = super
+          revenue += connection_bonus(routes.first.corporation) if routes.any?
```
Comment (tobymao) [inline lib/engine/game/g_18_ny/game.rb:437]: unless empty

File: lib/engine/game/g_18_ny/step/route.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+require_relative '../../../step/route'
+
+module Engine
+  module Game
+    module G18NY
+      module Step
+        class Route < Engine::Step::Route
+          def process_run_routes(action)
+            super
+            action.routes.map(&:stops).flatten.select { |stop| stop.hex.assigned?('connection_bonus') }.each do |stop|
```
Comment (tobymao) [inline lib/engine/game/g_18_ny/step/route.rb:12]: flat_map(&:stops).each do |stop|
   next if stop..

File: lib/engine/game/g_18_ny/game.rb (hunk 1 above)
Comment (crericha) [reply lib/engine/game/g_18_ny/game.rb:437]: Okay. I thought we determined performance was the same though?

Comment (tobymao) [reply lib/engine/game/g_18_ny/game.rb:437]: performance is not the same

Comment (crericha) [reply lib/engine/game/g_18_ny/game.rb]: Hmm...I thought Jen ran it both ways and determined it was. Maybe I'm misremembering.

**Pull request merged**